dbms/src/Encryption/AESCTRCipher.cpp

```diff
@@ -0,0 +1,280 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Common/TiFlashException.h>
+#include <Encryption/AESCTRCipher.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+
+#include <ext/scope_guard.h>
+
+
+namespace DB::Encryption
+{
+
+size_t keySize(EncryptionMethod method)
+{
+    switch (method)
+    {
+    case EncryptionMethod::Aes128Ctr:
+        return 16;
+    case EncryptionMethod::Aes192Ctr:
+        return 24;
+    case EncryptionMethod::Aes256Ctr:
+        return 32;
+    case EncryptionMethod::SM4Ctr:
+        return 16;
+    default:
+        return 0;
+    }
+}
+
+size_t blockSize(EncryptionMethod method)
+{
+    switch (method)
+    {
+    case EncryptionMethod::Aes128Ctr:
+    case EncryptionMethod::Aes192Ctr:
+    case EncryptionMethod::Aes256Ctr:
+        return AES_BLOCK_SIZE;
+    case EncryptionMethod::SM4Ctr:
+#if USE_GM_SSL
+        // SM4 uses 128-bit block size as AES.
+        // Ref: https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/include/crypto/sm4.h#L24
+        return 16;
+#elif OPENSSL_VERSION_NUMBER < 0x1010100fL || defined(OPENSSL_NO_SM4)
+        throw DB::TiFlashException(
+            Errors::Encryption::Internal,
+            "Do not support encryption method SM4Ctr when using OpenSSL and the version is less than 1.1.1");
+#else
+        // Openssl support SM4 after 1.1.1 release version.
+        return SM4_BLOCK_SIZE;
+#endif
```

Comment:
why not keep the previous definition?
```
#if defined(SM4_BLOCK_SIZE)
        return SM4_BLOCK_SIZE;
#else
        throw DB::TiFlashException(...
#endif
```

Reply:
```
#if (USE_GM_SSL == 0) && !defined(OPENSSL_NO_SM4)
// TODO: OpenSSL Lib does not export SM4_BLOCK_SIZE by now.
// Need to remove SM4_BLOCK_SIZE once Openssl lib support the definition.
// SM4 uses 128-bit block size as AES.
// Ref:
// https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/include/crypto/sm4.h#L24
#define SM4_BLOCK_SIZE 16
#endif
```

combine with this, make it easy to understand